### PR TITLE
Add GitHub Actions workflow

### LINKED ISSUE #4789: [Dev Process] Travis has stopped reporting build statuses to this repo
<details><summary>Travis used to report build statuses until July 16 or 17, 2020</summary>

screenshot of recent commits

</details>

What I expect: build status reported as green ✔️ or red ❌  marks next to each commit I push to an open PR, and a dedicated check status line at the bottom of PRs (where GitHub reports "This branch has no conflicts with the base branch" and the like). Without these, it's not really discoverable that CI checks are available.

### Turn it off and on again

[One person on Travis's support community reported that it worked](https://travis-ci.community/t/github-ci-notifiactions-from-travis-stopped-working/9238/4).

>  I just switched my repo off and on again in the Travis settings and started a rebuild, and it showed up in GitHub again, so it is worth a try.

https://travis-ci.org/account/repositories lets you turn a repo off and on.

Some recent builds to retry: [push to master](https://travis-ci.org/github/Jermolene/TiddlyWiki5/builds/714182778), [pull request update](https://github.com/Jermolene/TiddlyWiki5/pull/4716).

### Migrate from Legacy GitHub Services to GitHub Apps

[Travis acknowledged](https://travis-ci.community/t/github-status-not-posted-on-commits-on-repositories-using-legacy-service-integration/7798) that there may be issues with status updates for repos that are using Legacy GitHub Services integration.

From further reading, it appears to be that repo builds hosted on **travis-ci.org** are considered legacy, and those that have been migrated to **travis-ci.com** use the newer GitHub Apps integration:

> On May 2nd, 2018 Travis CI announced that open source projects will be joining private projects on travis-ci.com!
> [..]
> We are implementing these changes with GitHub Apps
> https://docs.travis-ci.com/user/migrate/open-source-on-travis-ci-com/

TiddlyWiki5's builds are hosted at travis-ci.org, so I think it uses the legacy integration.

The URL above explains how the migration steps look like on the travis-ci.org side, and [this article](https://docs.travis-ci.com/user/migrate/open-source-repository-migration) explains how to do it on the travis-ci.com side. I followed the latter for my own repos and it seemed to work.

Hosting of public repos on travis-ci.com is still considered to be in beta, so depending on TidddlyWiki5's policy, this might not yet be a desirable change.

### PR DESCRIPTION
**Pull request by ento**
Fixes #4789 (indirectly)

Adds `.github/workflows/ci.yml` that runs the same scripts as `.travis.yml` in three stages (jobs in GitHub Actions parlance): test, build-prerelease, build-tiddlywiki-com.

How it looks like in action (pun intended): https://github.com/ento/TiddlyWiki5/actions/runs/201083222

^This run is for a branch that contains [a commit](https://github.com/ento/TiddlyWiki5/commit/0493f641fb3547d584852ac88f948fa9bff453a6) that allows the `build-*` jobs to run on any branch, so that we can see how these jobs get run. The `build-*` jobs ultimately fail because the necessary push token is not configured.

What happens when this PR is merged without disabling Travis:

- The `test` job will get run on every PR and on every push to `master` and `tiddlywiki-com` branches
- The `build-prerelease` job will get run on every push to `master`, but will fail at the final `git push` command until a `GITHUB_PUSH_TOKEN` secret is added to the repo that has push access to the destination repo
- The `build-tiddlywiki-com` job will get run on every push to `tiddlywiki-com`, but will fail at the final `git push` command until a `GITHUB_PUSH_TOKEN` secret is added to the repo that has push access to the destination repo
- Travis builds will continue to run as well.
  - If  `GITHUB_PUSH_TOKEN` secret is configured and the `build-*` stages/jobs get run in close parallel between Travis and GitHub, I expect one of the runners to fail; first one to run `git push` will succeed and the other one won't be able to push due to their local branch being behind the remote. If builds for one of the services get too delayed, the sites can get reverted to an older state when the delayed builds finally get run.

## Further actions required to complete the switch from Travis

* [Write access needed] Test if the `build-prerelease` and `build-tiddlywiki-com` jobs of the workflow work
  * Create a test branch and add it to `on.push.branches` and adjust the `if` conditions of jobs so that they get run on that branch.  (This needs to be done in a branch in this repo, so that the workflow can access the secret configured in the next step.)
  * [Admin access needed] [Create a secret](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `GITHUB_PUSH_TOKEN` that has push access to `Jermolene/jermolene.github.io`
    * As with Travis's secrets, "secrets are not passed to the runner when a workflow is triggered from a forked repository." (From [the doc above](https://docs.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#using-encrypted-secrets-in-a-workflow))
  * Push to the test branch and ensure the build jobs work
  * If any adjustments to `ci.yml` were needed, undo the changes to `on.push.branches` and `if` conditions and merge the rest of the changes.
* Disable Travis (might want to do this first before configuring the secret, to avoid the race condition described above)
* Rename variables / simplify logic in `bin/ci-*` scripts and `ci.yml` that mention Travis

### CODE REVIEW

File: bin/ci-push.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+# Push output back to GitHub
+
+
+cd output || exit 1
+
+git config --global user.email "${COMMIT_EMAIL:-travis@travis-ci.org}" || exit 1
```
Comment (pmario) [inline bin/ci-push.sh:8]: you should probably also change the email and commit name info?

Comment (ento) [reply bin/ci-push.sh:8]: I wasn't sure how we'll want to go about doing the switch, so this script was left in a state where it's possible to run from both Travis and GitHub Actions.

In GitHub Actions, they'll be overridden by these `COMMIT_*` environment variables: https://github.com/Jermolene/TiddlyWiki5/pull/4793/files#diff-e9f950f17198d3d5e3122a44230a09b9R8-R10

[Travis doesn't set these `COMMIT_*` environment variables](https://docs.travis-ci.com/user/environment-variables#default-environment-variables) (just checked now, phew), so the string after `:-` will be used instead.

**Comment by ento** [comment]
I'm not sure why the workflow isn't getting triggered (the `on: pull_request:` part should make it so that a run gets triggered when PRs are [opened/reopened or pushed to](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request)). Maybe Actions need to be enabled for the repo??

**Comment by ento** [comment]
> Maybe Actions need to be enabled for the repo??

If this is the case, I'd expect some messaging to that effect in https://github.com/Jermolene/TiddlyWiki5/actions which can only be seen by repo admins.

Another possibility is that a PR from a forked repository cannot add a new workflow and has to be merged before it takes effect. If so, this PR could be merged without verifying the workflow in this repo, or the branch could be re-pushed to this repo and a new PR could be opened to supersede this one.

**Comment by rmunn** [comment]
> Another possibility is that a PR from a forked repository cannot add a new workflow and has to be merged before it takes effect. If so, this PR could be merged without verifying the workflow in this repo, or the branch could be re-pushed to this repo and a new PR could be opened to supersede this one.

I think that's it: I created a new workflow in a branch on my https://github.com/rmunn/Testing repo and it ran even before I'd created a PR for it. That repo has been forked by other people, and when they created a branch in their own repos without submitting a PR into my repo I didn't see an Action run in my repo.

So to watch this run in the main TiddlyWiki repo, @Jermolene (or someone else with write access to this repo, maybe @twMat) will need to fetch your [add-github-actions-workflow](https://github.com/ento/TiddlyWiki5/tree/add-github-actions-workflow) branch and push that branch to the main repo. Then this PR should trigger the action to run, though possibly you'd need to close and reopen the PR (or create a new commit in it) to trigger the run the first time.

**Comment by ento** [comment]
>  [..] push that branch to the main repo. Then this PR should trigger the action to run, though possibly you'd need to close and reopen the PR

It seems like this'll be the case! GitHub's [documentation](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#about-workflow-events) details how it looks up workflow files:

> 1. An event occurs on your repository, and the resulting event webhook has an associated commit SHA and Git ref.
> 2. The `.github/workflows` directory in your repository is searched for workflow files at the associated commit SHA or Git ref. The workflow files must be present in that commit SHA or Git ref to be considered.

Since this PR's commit SHA only lives in _my fork_ of the repo at the moment, step 2 will come up empty. Pushing this PR's branch to the main repo should make the new workflow file visible to GitHub Actions.

This means...if a fix needs to be pushed to my branch, then the new commit will need to be pushed to the main repo again. I get the hunch that making changes to workflow files from a forked repo isn't really expected.

**Comment by Jermolene** [comment]
Hi @ento @rmunn @pmario apologies for the delay. Is this ready to merge?

**Comment by pmario** [comment]
Hi, I don't have experience with github actions. So @ento and @rmunn need to be sure. ... 

@Jermolene You'll probably need to merge it and then one of them will need to create a PR, so we will see if it works. .. I think some more testing is needed **after** the merge.

**Comment by ento** [comment]
As the only thing the workflow can do currently is to run tests, I think it's safe to merge as-is, granted that it's okay to have red X's on the branches that push to the static site. (The stages to push will get run but will fail due to missing credentials until a token is configured.) I can comment these parts out if that's undesirable.

There will also be email/web notifications when the workflow fails, which [can be disabled on the user side](https://docs.github.com/en/github/managing-subscriptions-and-notifications-on-github/configuring-notifications#github-actions-notification-options).

> You'll probably need to merge it and then one of them will need to create a PR, so we will see if it works. .. I think some more testing is needed **after** the merge.

Yup, we haven't been able to actually test the workflow in this repo.

When this PR is merged, I can extract "Further actions required to complete the switch from Travis" in the PR description into a new issue to keep track of these next steps.

**Comment by Jermolene** [comment]
Thanks @ento 

> The stages to push will get run but will fail due to missing credentials until a token is configured

Does the code push to jermolene.github.io? Presumably I'll have to disable the Travis build process before enabling it.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v1
+        with:
+          node-version: "${{ env.NODE_VERSION }}"
+      - run: "./bin/test.sh"
+  build-prelease:
```
Comment (Jermolene) [inline .github/workflows/ci.yml:21]: "prelease" is a typo

**Review by Jermolene — COMMENTED** [review]
I just noticed a small typo

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ento) [reply .github/workflows/ci.yml:21]: Nice catch

Comment (ento) [inline .github/workflows/ci.yml:21]: Comment with a suggestion block to fix the prelease typo
```suggestion
  build-prerelease:
```

**Comment by ento** [comment]
> Does the code push to jermolene.github.io?

Yes, it uses the same script as Travis builds that pushes to jermolene.github.io.

>  Presumably I'll have to disable the Travis build process before enabling it.

Disabling Travis first before configuring the token would be the safest, I agree. I believe there should be a toggle in [Travis's account settings](https://travis-ci.org/account/repositories).

From our investigation of how GitHub Actions work, it seems most efficient if the switch-over can be handled by someone with write access to this repo, and the token for pushing to jermolene.github.io will need to be configured by @Jermolene.

**Comment by Jermolene** [comment]
Thanks @ento. I'm delighted to be moving away from Travis CI!

**Comment by pmario** [comment]
@ento ... I would like to make the TW Actions more powerful. ... 

# TLDR; What I would like to have as an end-result is:

 - I create a new "feature branch" in **my** TiddlyWiki5 repository
    - eg: "test-this-and-that-ACTION"

If the ACTION part is missing **stop the CI/CD** because it's a "normal" branch. 
If the ACTION part is there go on!
 
 - The "test" action should be run 
 - A "build-2-test-files" action should be run

The "build-2-test-files" should create 2 tw-files 

 - `empty-<date-time>.html` .... empty TW version
 - `content-<date-time>.html` .... tw5.com version with a lot of content to test with.

Those 2 files should be ZIPed into `<feature-branch-name>-<commit-hash>.zip`

The zip file _plus_ some automatic info should be added to an issue in MY repository. This issue can be used to discuss the new feature prior to creating a PR.

**Is this possible?**

----------------------

As I can see, at the moment most action elements are hardcoded, so they only work for Jeremy. 

I want, that the actions also run on "contributor repositories". So If we push to a branch, a "test-build" should be activated, as shown above. I think this will help Jeremy a lot.

In https://github.com/Jermolene/TiddlyWiki5/discussions/5484 Jeremy tells us, that he needs 20-30 minutes to merge a PR. .. I think we can reduce this time quite a bit. 

I know, that "loading and building" a TEST-TW from other branches takes some time. ... Preparing the "test scenario", that covers _all_ of the code changes, needs a lot of time. ... BUT the  initial "test scenario" can be created by the contributor. ... This will also make the intentions much clearer.

I will have a closer look at the "actions configuration" and the "build scripts" involved. ... But it would be nice, if you could help, since you already know, how to build a "test environment", that can be used to develop the new action scripts. 

It would be nice, if you could point me in the right direction.

**Comment by Jermolene** [comment]
Hi @pmario I think you're suggesting that (a) we start making a build of incoming PRs and (b) we distribute the build by attaching a ZIP to the original PR.

The reason I've stayed away from (a) in the past is the (small) chance that a malicious contributor could make a PR with spammy links or other inappropriate content that would then be published to tiddlywiki.com without further user intervention. Your suggestion of publishing the builds as ZIPs would perhaps be a good way to get around that issue.

**Comment by pmario** [comment]
@Jermolene I want to store the ZIP in an issue in the contributor repo. Not in your repo and not reachable from tiddlywiki.com. 

I want to use variables or other placeholders in the action scripts, that will allow the scripts to run everywhere.
